database/postgres/dal/refresh.py

```diff
@@ -0,0 +1,121 @@
+from sqlalchemy.orm import Session
+from sqlalchemy.exc import SQLAlchemyError
+from typing import Optional
+from datetime import datetime
+
+from database.models.refresh_token_tracking import RefreshTokenTracking
+from database.models.token_blocklist import TokenBlocklist
+from database.postgres.dto.refresh import refreshDTO
+
+class RefreshTokenDAL:
+    def __init__(self, db_session: Session):
+        self.db_session = db_session
+
+    def save_refresh_token(self, refresh_dto: refreshDTO, last_ip: str, last_device: str, nonce: str) -> Optional[int]:
+        try:
+            refresh_entry = None
+            
+            if refresh_dto.id:
+                refresh_entry = self.get_refresh_token_by_id(refresh_dto.id)
+
+            if refresh_entry:
+                refresh_entry.user_id = refresh_dto.user_id
+                refresh_entry.last_ip = last_ip
+                refresh_entry.last_device = last_device
+                refresh_entry.nonce = refresh_dto.nonce
+            else:
+                refresh_entry = RefreshTokenTracking(
+                    user_id=refresh_dto.user_id,
+                    last_ip=last_ip,
+                    last_device=last_device,
+                    nonce=nonce
+                )
+                self.db_session.add(refresh_entry)
+
+            self.db_session.commit()
+            self.db_session.refresh(refresh_entry)
+            return refresh_entry.id
+        except SQLAlchemyError as e:
+            self.db_session.rollback()
+            print(f"Error in save_refresh_token: {e}")
+            return None
+
+
+    def get_refresh_token_by_id(self, refresh_id: int) -> Optional[RefreshTokenTracking]:
+        return self.db_session.query(RefreshTokenTracking).filter(RefreshTokenTracking.id == refresh_id).first()
+
+    def get_valid_refresh_token_by_user(self, user_id: int) -> Optional[RefreshTokenTracking]:
+        return (
+            self.db_session.query(RefreshTokenTracking)
+            .filter(
+                RefreshTokenTracking.user_id == user_id,
+                RefreshTokenTracking.refresh_token.has(revoked=False),  
+                TokenBlocklist.expires_at > datetime.utcnow()
+            )
+            .first()
+        )
+
+    def get_nonce_by_user_id(self, user_id: int) -> Optional[str]:
+        entry = self.db_session.query(RefreshTokenTracking).filter(RefreshTokenTracking.user_id == user_id).first()
+        return entry.nonce if entry else None
```

Comment:
If you only need nonce, why selecting all the other fields? Select nonce only and then just `return nonce`